secant/Util/UserPreferencesStorage.swift

```diff
@@ -0,0 +1,115 @@
+//
+//  UserPreferencesStorage.swift
+//  secant-testnet
+//
+//  Created by Lukáš Korba on 03/18/2022.
+//
+
+import Foundation
+
+/// Representation of the user preferences stored in the local persistent storage (non-encrypted, no security needed)
+protocol UserPreferences {
```

Comment:
Instead of a protocol based approach, have you considered extending UesrDefaults to work with your own `Key` type that defines the name and the Value Type that is returned (and maybe a default?), so that we can strongly type and build the keys we need but use the regular UserDefaults as the dependency?

One downside to the current approach that I see is that UserDefaults has many capabilities, and a larger API surface than just setting and retrieving values.  If we need any of that functionality at all, it will need to be exposed through this protocol because the underlying UserDefaults is privately held, or you'll have to work with the underlying source of truth external to this protocol.

Another factor here is that when you define a new key with this approach, it needs to be added to the protocol, added to the initializer for a default, and then implemented with a custom getter and setter for every key.

If we instead extend UserDefaults to work with strongly typed keys that we define, we could stick with a more vanilla dependency, reduce some of the duplications of data and allow us to work directly with UserDefaults.

From a testing perspective, you can setup static initializers on UserDefaults to put yourself in any state that you like, without needing a protocol.

https://www.swiftbysundell.com/articles/the-power-of-userdefaults-in-swift/
Some ideas here to reference if you want to consider this approach.

Reply:
>Instead of a protocol based approach, have you considered extending UesrDefaults to work with your own Key type that defines the name and the Value Type that is returned (and maybe a default?), so that we can strongly type and build the keys we need but use the regular UserDefaults as the dependency?

This sounds like a good idea to explore but I can't wrap my head around it. It sounds like a whole re-write though.

are you referring to so something like

````Swift
struct SettingKey<T: Any>: Hashable {
    associatedType T
    var key: String
    var value: T
}

extension UserDefaults {
    func getValue<K, T>(for key: K) -> T where K: SettingsKey, T: Settings.T {}
    func setValue<K, T>(_ value: T, for key: T) where K: SettingsKey, T: Settings.T {}
}
````

note: I wrote this on the comment editor, I'm not even sure if this is possible.